File: Others/Practicals/Practicals/quick_sort.cpp
#include <iostream>

using namespace std;

void swap(int *a, int *b){
    int temp = *a;
    *a = *b;
    *b = temp;
}

int partition(int arr[], int low, int high){
    int pivot = arr[high];
    
    int i = low - 1;
    
    for(int j = low; j <= high - 1; j++){
        if(arr[j] < pivot){
            i++;
            swap(&arr[i], &arr[j]);
        }
        
    }
    swap(&arr[i + 1], &arr[high]);
    return i + 1;
}

void quicksort(int arr[], int low, int high){
    if(low < high){
        int pi = partition(arr, low, high);
        
        quicksort(arr, low, pi - 1);
        quicksort(arr, pi + 1, high);
    }
}

void printArray(int arr[], int size) 
{ 
    int i; 
    for (i = 0; i < size; i++) 
        cout << arr[i] << " "; 
    cout << endl; 
} 
  
int main() 
{ 
    int arr[] = {1915332, 1915321, 1915365, 1915335, 1915369, 1915375, 1915309, 1915384};
    int n = sizeof(arr) / sizeof(arr[0]); 
    quicksort(arr, 0, n - 1); 
    cout << "Sorted array: \n"; 
    printArray(arr, n); 
    return 0; 
} 

File: Javascript_DSALGO/Sort Algorithms/QuickSort.js
// Quick sort
function partition(arr, startIndex = 0, endIndex = arr.length - 1) {
  const swap = (arr, idx1, idx2) => {
    [arr[idx1], arr[idx2]] = [arr[idx2], arr[idx1]];
  };

  let pivot = arr[startIndex],
    pvtIdx = startIndex;

  for (let i = startIndex + 1; i <= endIndex; i++) {
    if (pivot > arr[i]) {
      pvtIdx++;
      swap(arr, pvtIdx, i);
    }
  }
  swap(arr, startIndex, pvtIdx);
  return pvtIdx;
}

function quickSort(arr, left = 0, right = arr.length - 1) {
  if (left < right) {
    let pvtIdx = partition(arr, left, right);

    quickSort(arr, left, pvtIdx - 1);
    quickSort(arr, pvtIdx + 1, right);
  }
  return arr;
}

// with Comparator
function pivot(arr, comp, startIndex = 0, endIndex = arr.length - 1) {
  const swap = (arr, idx1, idx2) => {
    [arr[idx1], arr[idx2]] = [arr[idx2], arr[idx1]];
  };

  let pivot = arr[startIndex],
    pvtIdx = startIndex;

  for (let i = startIndex + 1; i <= endIndex; i++) {
    if (comp(pivot, arr[i]) >= 0) {
      pvtIdx++;
      swap(arr, pvtIdx, i);
    }
  }
  swap(arr, startIndex, pvtIdx);
  return pvtIdx;
}

function quickSort(arr, comp, left = 0, right = arr.length - 1) {
  if (!comp) comp = (a, b) => a - b;
  if (left < right) {
    let pvtIdx = pivot(arr, comp, left, right);

    quickSort(arr, comp, left, pvtIdx - 1);
    quickSort(arr, comp, pvtIdx + 1, right);
  }
  return arr;
}
